### ISSUE TITLE
Document the Panache stream() operations for the need of transaction

### ISSUE THREAD

**Issue by loicmathieu** [quarkusio/quarkus#3256]
**Description**
Following the discussion on the mailing list (https://groups.google.com/d/msg/quarkus-dev/01GSKgzQ9dA/GjgE2c7nAwAJ) I propose to add to the Panache guide and to the Javadoc of the stream() operation that it can needs a transaction.

**Implementation ideas**
Add a note to the Panache guide
Add a not in the Javadoc

**Comment by loicmathieu** [comment]
I will propose a PR for this.

**Comment by loicmathieu** [comment]
To avoid multiple push and as documentation is always tricky, I propose in a comment the two addons that I propose to add to the documentation.

First, the Javadoc part :
```
   Find entities matching a query, with optional indexed parameters.
   This method is a shortcut for <code>find(query, params).stream()</code>.
    + As it uses a database cursor, a transaction may be needed to keep the cursor opens 
    + until the end of the processing of the stream. 
    + This is needed when one of the stream processing steps access the database.
```

Then, le Panache guide: I plan to add a note to the section of the stream method:
```
NOTE: The `stream` methods uses a database cursor so it may need a transaction to works properly. It is in fact 
needed when one of the stream steps make an access to the database. In this case the `stream` methods must be called
inside a method annotated by the `@Transactional` annotation.
```

@Sanne what do you think of my proposal ?

**Comment by Sanne** [comment]
The javadoc looks good in principle.

When you say "a transaction may be needed" .. I guess it depends on how the other issue gets resolved, but I really think that we'll end up saying that "a transaction is required". It also helps keeping things simple for users ;)

Some minor typos:

s/keep the cursor opens/keep the cursor open/
s/the end of the processing of the stream/the end of processing of the stream/

Regarding the Panache guide; I would keep it simple as in "the `stream` method requires a transaction" ?

**Comment by FroMage** [comment]
+1

**Comment by loicmathieu** [comment]
v2:

JavaDoc:
```
   Find entities matching a query, with optional indexed parameters.
   This method is a shortcut for <code>find(query, params).stream()</code>.
    + As it uses a database cursor, a transaction is needed to keep the cursor open 
    + until the end of the processing of the stream. 
    + Otherwise it will generate an error when one of the stream processing steps access the database.
```

Panache guide:
```
NOTE: The `stream` methods uses a database cursor, so it needs a transaction to works properly. This means that the `stream` methods must be called inside a method annotated by the `@Transactional` annotation.
```

I keep some more explaination as I like to have the _why_ in the documentation but I agree it's more straighforward to just include a note like:
```
NOTE: The `stream` methods requires a transaction to works properly.
```

**Comment by Sanne** [comment]
typo:
s/processing steps access the database/processing steps accesses the database/

But also I don't think the explanation is comprehensive; IMO if you really want to explain the reasons, it should be clear: it's not only required in case any subsequent stream processing step requires to load (modify) anything in the DB, it's also necessary simply to load the next elements of the stream.

s/transaction to works properly/transaction to work properly/

**Comment by Sanne** [comment]
> as we discussed on the mailing list, it's not mandatory to get the next element of the stream

no, sorry if I wasn't clear, but as I also said on the mailing list, this is actually mandatory for many databases: you won't be able to fetch the next item from the stream in lack of a transaction.
I gave you a direct link to the postgres documentation.

Not sure which database you're testing, but if you want a general API we need to satisfy the requirements of all major databases.

**Comment by Sanne** [comment]
@loicmathieu I know how Hibernate fetching works ;) But I'm telling you that it's not always going to fetch all results at once; there's several options to actually really stream the data, so it's still going to be required - especially when it actually makes sense to use a _stream_ . If not, people could just use _list_ , as I mentioned on the mailing list.

**Comment by loicmathieu** [comment]
@Sanne it wasn't my intention to explain you how hibernate work :)

Just saying that is works in some cases so it's not intuitive that a transaction is needed.

Anyway, OK for the v3 wordings ?

**Comment by Sanne** [comment]
yes the v3 wording looks ok to me.  I'm not entirely convinced by "It requires a transaction to works properly" as people will be wondering what "properly" means exactly.. but if you like that better, go for it.

N.B. that if we decide to check for the transaction to be mandatory "properly" will mean "without throwing an exception" so I guess we'll want to amend the message if it comes to that :)

**Comment by loicmathieu** [comment]
@Sanne documentation done.
I also reference this issue in the one that will investiguate the enforcement of a transaction for those methods so if we go this path we don't forget to update the documentation path.

**Comment by loicmathieu** [comment]
@Sanne do you validate the documentation?

**Comment by loicmathieu** [comment]
Fixed by #3335
